### PR TITLE
Search with combined indices

### PR DESCRIPTION
**Pull request by zzb111**

### CODE REVIEW

**Comment by harsha-simhadri** [comment]
> _No description provided._

Thanks for the PR. 
Please add some test cases to algos.yaml and leave a message here when you think the algo is ready for testing

**Comment by harsha-simhadri** [comment]
@zzb111  On the public query set, here are the results I see for deep-1B. Is this as expected? Can you truncate the number of query arg combinations to 10?

team11,"FaissIVFPQ(nprobe=16,quantizer_efSearch=128)",deep-1B,10,23142.48147328683,19134.4202,1000000.0,43995212.0,1901.0585382031438,0,0,0.55932
team11,"FaissIVFPQ(nprobe=32,quantizer_efSearch=64)",deep-1B,10,20842.049190429054,38190.0385,1000000.0,43995212.0,2110.8870628807067,0,0,0.6119399999999999
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=128)",deep-1B,10,10703.891937828013,75897.8232,1000000.0,43995212.0,4110.207040162563,0,0,0.64955
team11,"FaissIVFPQ(nprobe=16,quantizer_efSearch=64)",deep-1B,10,31144.033312938463,19150.0692,1000000.0,43995212.0,1412.6369426185609,0,0,0.55869
team11,"FaissIVFPQ(nprobe=32,quantizer_efSearch=16)",deep-1B,10,25549.8017505889,38367.5681,1000000.0,43995212.0,1721.9394666726112,0,0,0.59177
team11,"FaissIVFPQ(nprobe=128,quantizer_efSearch=64)",deep-1B,10,6285.34538721455,151037.2369,1000000.0,43995212.0,6999.649070915603,0,0,0.66567
team11,"FaissIVFPQ(nprobe=8,quantizer_efSearch=64)",deep-1B,10,45529.73601397273,9590.338,1000000.0,43995212.0,966.2962242192268,0,0,0.48758999999999997
team11,"FaissIVFPQ(nprobe=2,quantizer_efSearch=4)",deep-1B,10,266557.1874344618,2432.805,1000000.0,43995212.0,165.04980572252276,0,0,0.2394
team11,"FaissIVFPQ(nprobe=8,quantizer_efSearch=8)",deep-1B,10,96432.08744013666,9698.7188,1000000.0,43995212.0,456.2300077483177,0,0,0.44147
team11,"FaissIVFPQ(nprobe=4,quantizer_efSearch=8)",deep-1B,10,150685.0752113354,4853.7693,1000000.0,43995212.0,291.9679466483116,0,0,0.37116
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=512)",deep-1B,10,6491.821192232357,75831.7416,1000000.0,43995212.0,6777.021531745434,0,0,0.6504800000000001
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=256)",deep-1B,10,8716.516372728966,75846.0382,1000000.0,43995212.0,5047.33888157959,0,0,0.65028
team11,"FaissIVFPQ(nprobe=2,quantizer_efSearch=8)",deep-1B,10,200392.92130107404,2427.8048,1000000.0,43995212.0,219.54474097366332,0,0,0.2879
team11,"FaissIVFPQ(nprobe=256,quantizer_efSearch=128)",deep-1B,10,2982.6414175522623,298677.2549,1000000.0,43995212.0,14750.419457430173,0,0,0.68065
team11,"FaissIVFPQ(nprobe=32,quantizer_efSearch=128)",deep-1B,10,16936.67821270829,38146.8733,1000000.0,43995212.0,2597.629325388527,0,0,0.61319
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=64)",deep-1B,10,12101.040112011618,76033.925,1000000.0,43995212.0,3635.655414143276,0,0,0.64679
team11,"FaissIVFPQ(nprobe=32,quantizer_efSearch=256)",deep-1B,10,12511.350826676657,38131.6387,1000000.0,43995212.0,3516.423814620686,0,0,0.6136
team11,"FaissIVFPQ(nprobe=128,quantizer_efSearch=512)",deep-1B,10,4480.126194960875,150518.7058,1000000.0,43995212.0,9820.083204237555,0,0,0.67159
team11,"FaissIVFPQ(nprobe=256,quantizer_efSearch=64)",deep-1B,10,3129.3109307622617,298407.7978,1000000.0,43995212.0,14059.073378586674,0,0,0.6748000000000001
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=16)",deep-1B,10,13194.693185847027,76182.142,1000000.0,43995212.0,3334.311103739071,0,0,0.61593
team11,"FaissIVFPQ(nprobe=16,quantizer_efSearch=8)",deep-1B,10,48728.36775284867,19307.0428,1000000.0,43995212.0,902.8665237289429,0,0,0.52105
team11,"FaissIVFPQ(nprobe=16,quantizer_efSearch=32)",deep-1B,10,38254.397933646105,19187.3264,1000000.0,43995212.0,1150.0693874809265,0,0,0.55522
team11,"FaissIVFPQ(nprobe=8,quantizer_efSearch=16)",deep-1B,10,81003.80269259219,9639.8788,1000000.0,43995212.0,543.1252674267769,0,0,0.47639
team11,"FaissIVFPQ(nprobe=128,quantizer_efSearch=128)",deep-1B,10,5989.443977410837,150749.528,1000000.0,43995212.0,7345.458470924473,0,0,0.67008
team11,"FaissIVFPQ(nprobe=128,quantizer_efSearch=32)",deep-1B,10,6456.809400457732,150956.0504,1000000.0,43995212.0,6813.769661046696,0,0,0.65384
team11,"FaissIVFPQ(nprobe=1,quantizer_efSearch=4)",deep-1B,10,342663.4968097187,1221.0374,1000000.0,43995212.0,128.3918841942787,0,0,0.17378
team11,"FaissIVFPQ(nprobe=8,quantizer_efSearch=4)",deep-1B,10,100202.44491534535,9713.4815,1000000.0,43995212.0,439.0632587574959,0,0,0.42394
team11,"FaissIVFPQ(nprobe=2,quantizer_efSearch=16)",deep-1B,10,144520.53944911138,2412.999,1000000.0,43995212.0,304.42186396274565,0,0,0.30196
team11,"FaissIVFPQ(nprobe=32,quantizer_efSearch=32)",deep-1B,10,24092.22160250486,38283.8878,1000000.0,43995212.0,1826.1168573771477,0,0,0.60582
team11,"FaissIVFPQ(nprobe=128,quantizer_efSearch=256)",deep-1B,10,5465.330567942726,150575.0368,1000000.0,43995212.0,8049.872089724446,0,0,0.67131
team11,"FaissIVFPQ(nprobe=64,quantizer_efSearch=32)",deep-1B,10,12767.846597253256,76197.9369,1000000.0,43995212.0,3445.781687999344,0,0,0.63872
team11,"FaissIVFPQ(nprobe=4,quantizer_efSearch=4)",deep-1B,10,179588.355434145,4867.3057,1000000.0,43995212.0,244.97808832674028,0,0,0.30093000000000003
team11,"FaissIVFPQ(nprobe=8,quantizer_efSearch=32)",deep-1B,10,63345.914001392484,9603.2599,1000000.0,43995212.0,694.5232805233956,0,0,0.48559
team11,"FaissIVFPQ(nprobe=16,quantizer_efSearch=16)",deep-1B,10,44720.63888089221,19260.403,1000000.0,43995212.0,983.7787004156113,0,0,0.54052
team11,"FaissIVFPQ(nprobe=4,quantizer_efSearch=16)",deep-1B,10,115404.72646329686,4826.7691,1000000.0,43995212.0,381.22539126672746,0,0,0.39386

**Comment by harsha-simhadri** [comment]
@zzb111  I see the following error when I try to run bigann-1B with your algorithm. Could you please investigate.

2021-11-04 09:59:37,822 - annb.abbf5ada41 - INFO - Trying to instantiate benchmark.algorithms.faiss_t1_plus.Faiss(['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}])
2021-11-04 09:59:38,202 - annb.abbf5ada41 - INFO - Running team11 on bigann-1B
2021-11-04 09:59:38,203 - annb.abbf5ada41 - INFO - Got 10000 queries
2021-11-04 09:59:38,205 - annb.abbf5ada41 - INFO - Build-time number of threads: 32
2021-11-04 09:59:38,210 - annb.abbf5ada41 - INFO - Traceback (most recent call last):
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -   File "run_algorithm.py", line 3, in <module>
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO - Update add-time parameters
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -    update quantizer efSearch= 16 -> 64
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -   update quantizer efConstruction= 40 -> 168
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO - setting maxtrain to 3276800
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO - getting first 3276800 dataset vectors for training
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -     run_from_cmdline()
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -   File "/home/app/benchmark/runner.py", line 227, in run_from_cmdline
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -     run(definition, args.dataset, args.count, args.runs, args.rebuild,
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -   File "/home/app/benchmark/runner.py", line 94, in run
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -     algo.fit(dataset)
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -   File "/home/app/benchmark/algorithms/faiss_t1_plus.py", line 190, in fit
2021-11-04 09:59:38,211 - annb.abbf5ada41 - INFO -     xt2 = next(ds.get_dataset_iterator(bs=maxtrain))
2021-11-04 09:59:38,212 - annb.abbf5ada41 - INFO -   File "/home/app/benchmark/datasets.py", line 295, in get_dataset_iterator
2021-11-04 09:59:38,212 - annb.abbf5ada41 - INFO -     filename = self.get_dataset_fn()
2021-11-04 09:59:38,212 - annb.abbf5ada41 - INFO -   File "/home/app/benchmark/datasets.py", line 290, in get_dataset_fn
2021-11-04 09:59:38,212 - annb.abbf5ada41 - INFO -     raise RuntimeError("file not found")
2021-11-04 09:59:38,212 - annb.abbf5ada41 - INFO - RuntimeError: file not found
2021-11-04 09:59:38,678 - annb.abbf5ada41 - ERROR - ['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}]
Trying to instantiate benchmark.algorithms.faiss_t1_plus.Faiss(['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}])
Running team11 on bigann-1B
Got 10000 queries
Build-time number of threads: 32
Traceback (most recent call last):
  File "run_algorithm.py", line 3, in <module>
Update add-time parameters
   update quantizer efSearch= 16 -> 64
  update quantizer efConstruction= 40 -> 168
setting maxtrain to 3276800
getting first 3276800 dataset vectors for training
    run_from_cmdline()
  File "/home/app/benchmark/runner.py", line 227, in run_from_cmdline
    run(definition, args.dataset, args.count, args.runs, args.rebuild,
  File "/home/app/benchmark/runner.py", line 94, in run
    algo.fit(dataset)
  File "/home/app/benchmark/algorithms/faiss_t1_plus.py", line 190, in fit
    xt2 = next(ds.get_dataset_iterator(bs=maxtrain))
  File "/home/app/benchmark/datasets.py", line 295, in get_dataset_iterator
    filename = self.get_dataset_fn()
  File "/home/app/benchmark/datasets.py", line 290, in get_dataset_fn
    raise RuntimeError("file not found")
RuntimeError: file not found

File: algos.yaml [hunk 1]
```diff
@@ -352,6 +471,47 @@ bigann-1B:
...
+      docker-tag: billion-scale-benchmark-faissconda
+      module: benchmark.algorithms.faiss_t1_plus
+      constructor: Faiss
+      base-args: ["@metric"]
+      run-groups:
+        base:
+          args: |
+              [{
+                "indexkey": "OPQ128_256,IVF65536_HNSW32,PQ128x4fsr",
+                "url": "https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex"
```
Comment (harsha-simhadri) [inline algos.yaml:484]: @zzb111  Can you please check that running bigann-1B downloads this index file and actually runs? It doesn't seem to work:

021-11-04 12:29:35,740 - annb.95019856d3 - INFO - Trying to instantiate benchmark.algorithms.faiss_t1_plus.Faiss(['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}])
2021-11-04 12:29:36,118 - annb.95019856d3 - INFO - Running team11 on bigann-1B
2021-11-04 12:29:36,118 - annb.95019856d3 - INFO - Got 10000 queries
2021-11-04 12:29:36,122 - annb.95019856d3 - INFO - Build-time number of threads: 32
2021-11-04 12:29:36,126 - annb.95019856d3 - INFO - Update add-time parameters
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -    update quantizer efSearch= 16 -> 64
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -   update quantizer efConstruction= 40 -> 168
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO - setting maxtrain to 3276800
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO - getting first 3276800 dataset vectors for training
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO - Traceback (most recent call last):
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -   File "run_algorithm.py", line 3, in <module>
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -     run_from_cmdline()
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -   File "/home/app/benchmark/runner.py", line 227, in run_from_cmdline
2021-11-04 12:29:36,127 - annb.95019856d3 - INFO -     run(definition, args.dataset, args.count, args.runs, args.rebuild,
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -   File "/home/app/benchmark/runner.py", line 94, in run
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -     algo.fit(dataset)
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -   File "/home/app/benchmark/algorithms/faiss_t1_plus.py", line 190, in fit
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -     xt2 = next(ds.get_dataset_iterator(bs=maxtrain))
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -   File "/home/app/benchmark/datasets.py", line 295, in get_dataset_iterator
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -     filename = self.get_dataset_fn()
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -   File "/home/app/benchmark/datasets.py", line 290, in get_dataset_fn
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO -     raise RuntimeError("file not found")
2021-11-04 12:29:36,128 - annb.95019856d3 - INFO - RuntimeError: file not found
2021-11-04 12:29:36,590 - annb.95019856d3 - ERROR - ['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}]
Trying to instantiate benchmark.algorithms.faiss_t1_plus.Faiss(['euclidean', {'indexkey': 'OPQ128_256,IVF65536_HNSW32,PQ128x4fsr', 'url': 'https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/bigann-1B.IVF1M_2level_PQ64x4fsr.faissindex'}])
Running team11 on bigann-1B
Got 10000 queries
Build-time number of threads: 32
Update add-time parameters
   update quantizer efSearch= 16 -> 64
  update quantizer efConstruction= 40 -> 168
setting maxtrain to 3276800
getting first 3276800 dataset vectors for training
Traceback (most recent call last):
  File "run_algorithm.py", line 3, in <module>
    run_from_cmdline()
  File "/home/app/benchmark/runner.py", line 227, in run_from_cmdline
    run(definition, args.dataset, args.count, args.runs, args.rebuild,
  File "/home/app/benchmark/runner.py", line 94, in run
    algo.fit(dataset)
  File "/home/app/benchmark/algorithms/faiss_t1_plus.py", line 190, in fit
    xt2 = next(ds.get_dataset_iterator(bs=maxtrain))
  File "/home/app/benchmark/datasets.py", line 295, in get_dataset_iterator
    filename = self.get_dataset_fn()
  File "/home/app/benchmark/datasets.py", line 290, in get_dataset_fn
    raise RuntimeError("file not found")
RuntimeError: file not found

File: algos.yaml [hunk 2]
```diff
@@ -169,6 +205,53 @@ deep-1B:
...
+      constructor: Faiss
+      base-args: ["@metric"]
+      run-groups:
+        base:
+          args: |
+              [{"indexkey": "OPQ128_256,IVF65536_HNSW32,PQ128x4fsr",
+                "url": "https://dl.fbaipublicfiles.com/billion-scale-ann-benchmarks/track1_baseline_faiss/deep-1B.IVF1M_2level_PQ64x4fsr.faissindex"
+                }]
+          query-args: |
+              ["nprobe=1,quantizer_efSearch=4",
```
[the remark this answers is not in the document]
Comment (zzb111) [reply algos.yaml]: OK, we already done.

File: algos.yaml [hunk 3]
```diff
@@ -169,6 +205,53 @@ deep-1B:
               "nprobe=128,quantizer_efSearch=512",
               "nprobe=256,quantizer_efSearch=64",
               "nprobe=256,quantizer_efSearch=128"]
+    team11:
+      docker-tag: billion-scale-benchmark-faissconda
+      module: benchmark.algorithms.faiss_t1
```
[the remark this answers is not in the document]
Comment (zzb111) [reply algos.yaml]: No, it's just a writing mistake and we have revised it.

**Pull request merged**